2024/BIN-2024-0003.md

```diff
@@ -0,0 +1,124 @@
+* For legacy and segwit v0 scripts, if the public key size is 33 bytes and its first byte is 0x02 or 0x03, it is considered a compressed public key as described in BIP137:
+    * If the signature is not the empty vector, the signature is validated against the public key and message using ECDSA. Validation failure in this case immediately terminates script execution with failure.
+* If the public key size is not zero, and it is not a BIP340 public key, nor a BIP137 compressed public key; the public key is of an unknown public key type, and no actual signature verification is applied. During script execution of signature opcodes they behave exactly as known public key types except that signature validation is considered to be successful.
+* If the script did not fail and terminate before this step, regardless of the public key type:
+    * If the signature is the empty vector:
+        * For `OP_CHECKSIGFROMSTACKVERIFY`, the script MUST fail and terminate immediately.
+        * For `OP_CHECKSIGFROMSTACK`, an empty vector is pushed onto the stack, and execution continues with the next opcode.
+    * If the signature is not the empty vector:
+        * For tapscript 0xc0, the opcode is counted towards the sigops budget as described in BIP341.
+        * For legacy and segwit v0, the opcode is counted towards the sigops limit, as described in BIP141
+        * For `OP_CHECKSIGFROMSTACKVERIFY`, execution continues without any further changes to the stack.
+        * For `OP_CHECKSIGFROMSTACK`, a 1-byte value 0x01 is pushed onto the stack.
+
+## Design Considerations
+
+1. Message hashing: ECDSA requires the message to be securely hashed to 32-bytes before being used in the signing protocol, so a SHA256 hash is taken before the message is passed for ECDSA signing. BIP340 is compatible with any size of message and does not require it to be a securely hashed input, so the message is not hashed prior to BIP340 signing.
+2. Verify NOP upgrade: To bring both ECDSA and BIP340 signing to bitcoin script, a NOP upgrade path was chosen for `OP_CHECKSIGFROMSTACKVERIFY`. This necessarily means leaving the 3 arguments on the stack when executing `OP_CHECKSIGFROMSTACKVERIFY`. Scripts will need to drop or otherwise manage these stack elements.
+3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK(VERIFY)`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD OP_TOALTSTACK`).
+4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for future bitcoin upgrades, such as `OP_CAT`.
+5. BIP118-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing OP_INTERNALKEY or separately introducing that key type for all Tapscript signature checking operations in a separate change.
```

Comment:
If you write the references to other bips as `[BIP 118]` and the references below as `[BIP 118]: https://...` you should get inline links without much more typing.